### ISSUE TITLE
Objective-C improvements

### ISSUE THREAD

**Issue by dvdsgl** [quicktype/quicktype#434]
Reading about [Mantle](https://github.com/Mantle/Mantle), a popular JSON framework for Objective-C, I spotted some good improvements for our Objective-C output:

- [ ] Support enums
- [ ] Support `NSCopying` and `NSCoding`
- [ ] Use proper `readonly`, `strong`, and `copy` property attributes

**Comment by dvdsgl** [comment]
It seems that exceptions may be out of fashion in Objective-C as well...

**Comment by dvdsgl** [comment]
`QTTopLevelFromData` should catch exceptions and write to the `NSError **error` so there's only one style of error handling there.

**Comment by n8chur** [comment]
I noticed that you can annotate a property as `nonnull` by adding the property to the list of `required` fields in a schema. Is this nullability validated in any way?

Ideally it would be considered an error if one attempted to parse a model object with a JSON payload whose `required` property is missing or `null`.

**Comment by n8chur** [comment]
I agree to exceptions being "out of fashion". It's more common to pass a pointer to an `NSError` and optionally have it populated if the method in question returns `NO` or `nil` (depending on context).

**Comment by n8chur** [comment]
Defining an `enum` on a property of type `string` does not seem to generate an enum in the generated model (even when the appropriate checkbox is checked).

**Comment by dvdsgl** [comment]
@n8chur because of our recursive marshaling of dynamic values, I think threading the `NSError` in and out of our property initializers is going to be way too cumbersome. My plan is to still use exceptions internally but to catch them at the top level and convert them to `NSError`.

**Comment by dvdsgl** [comment]
@n8chur `nonnull` is validated in the `NSDictionary objectForKey:withClass:` method--this is also the source of the exceptions design choice.

**Comment by dvdsgl** [comment]
Our dynamic checking is far from comprehensive, though. For example, collections of boxed primitives (e.g. `NSArray<NSNumber *> *`) do not get their elements checked.

**Comment by dvdsgl** [comment]
@n8chur `enum` is still on our todo list (above), we just need to decide how to implement it. What would you expect/desire?

**Comment by n8chur** [comment]
@dvdsgl It would be great to have properties with typed enums generated something like:
```
typedef NS_ENUM(NSInteger, SomeEnum){
    SomeEnumValueA,
    SomeEnumValueB
};
```

Then the property might look like:
```
@property (readonly, nonatomic) SomeEnum enumValue;
```
All of the translation between `NSString` and the enum values could be under the hood and the strings themselves do not necessarily need to be exposed publicly.

**Comment by dvdsgl** [comment]
@n8chur excellent, that's exactly what we had in mind. I should get it in the next couple of days, but feel free to send a PR if you're feeling ambitious 😄 We could really use some passionate contributors with more specialized per-language knowledge.

**Comment by dvdsgl** [comment]
@n8chur alright, speedbump – what would you expect to happen when `enums` appear as values in JSON arrays and maps? Would you expect an `NSArray<NSNumber *>`? That seems a bit... tragic. [Here's a sample](https://gist.github.com/dvdsgl/cf2df4ed93ebf73fb527cd51a217ac15#file-qtpokedex-m) and [where we get into trouble](https://gist.github.com/dvdsgl/cf2df4ed93ebf73fb527cd51a217ac15#file-qtpokedex-m-L90).

**Comment by n8chur** [comment]
@dvdsgl I'm currently working on open sourcing a code/test fixture generation tool at [Automatic](https://github.com/automatic) but I may dive in after that's wrapped up!

**Comment by n8chur** [comment]
@dvdsgl Regarding enums: `NSArray<NSNumber *> *` seems reasonable, but perhaps with a comment about what that number is.

Another thing to consider is to use a box class for enum values when they need to be contained in `NSArray`s/`NSDictionary`s.

Something like:
```
@interface QTWeaknessBox : NSObject

@property (readonly, nonatomic) QTWeakness value;

@end

...

@property (nonatomic, readonly, copy) NSArray<QTWeaknessBox *> *weaknesses;
```

**Comment by dvdsgl** [comment]
@n8chur alright... Have a look: https://app.quicktype.io/#l=objc&s=pokedex

I had to make 'pseudo-enums' for reference semantics (boxing) and to preserve readable type names when enums become type parameters (e.g. I want `NSArray<QTWeakness *>` not `NSArray<NSNumber *>`):

```objc
@interface QTWeakness : NSObject
@property (nonatomic, readonly, copy) NSString *value;
+ (instancetype _Nullable)withValue:(NSString *)value;
+ (QTWeakness *)bug;
+ (QTWeakness *)dark;
+ (QTWeakness *)dragon;
+ (QTWeakness *)electric;
@end

@implementation QTWeakness
static NSMutableDictionary<NSString *, QTWeakness *> *qtWeaknesses;
@synthesize value;

+ (QTWeakness *)bug { return qtWeaknesses[@"Bug"]; }
+ (QTWeakness *)dark { return qtWeaknesses[@"Dark"]; }
+ (QTWeakness *)dragon { return qtWeaknesses[@"Dragon"]; }
+ (QTWeakness *)electric { return qtWeaknesses[@"Electric"]; }

+ (void)initialize
{
    NSArray<NSString *> *values = @[
        @"Bug",
        @"Dark",
        @"Dragon",
        @"Electric",
    ];
    qtWeaknesses = [NSMutableDictionary dictionaryWithCapacity:values.count];
    for (NSString *value in values) qtWeaknesses[value] = [[QTWeakness alloc] initWithValue:value];
}

+ (instancetype _Nullable)withValue:(NSString *)value { return qtWeaknesses[value]; }

- (instancetype)initWithValue:(NSString *)val
{
    if (self = [super init]) value = val;
    return self;
}

- (NSUInteger)hash { return value.hash; }
@end
```

Perhaps this is a tad unorthodox, but I think it's the best for now, on balance.

You can see in the code that I track the set of these pseudo-enums, which means they can be differentiated from natural `NSUInteger` enums. It makes sense to emit natural enums if they aren't used within generic types, so this leaves room for improvement. In the `pokedex.json` sample, `Pokemon.egg` could be emitted as a natural enum, for example.

**Comment by schani** [comment]
We do not take feature requests anymore. Please consider making a pull request instead. We encourage the use of coding agents, but please use at least an Opus-class model.
